docs/features/incremental-generators.cookbook.md

```diff
@@ -0,0 +1,810 @@
+
+### Generated class
+
+**User scenario:** As a generator author I want to be able to add a type to the compilation, that can be referenced by the user's code. Common use cases include
+creating an attribute that will be used to drive other source generator steps.
+
+**Solution:** Have the user write the code as if the type was already present. Generate the missing type based on information available in the compilation using
+the `RegisterPostInitializationOutput` step.
+
+**Example:**
+
+Given the following user code:
+
+```csharp
+public partial class UserClass
+{
+    [GeneratedNamespace.GeneratedAttribute]
+    public partial void UserMethod();
+}
+```
+
+Create a generator that will create the missing type when run:
+
+```csharp
+[Generator]
+public class CustomGenerator : IIncrementalGenerator
+{
+    public void Initialize(IncrementalGeneratorInitializationContext context)
+    {
+        context.RegisterPostInitializationOutput(static postInitializationContext => {
+            postInitializationContext.AddSource("myGeneratedFile.cs", SourceText.From("""
+                namespace GeneratedNamespace
+                {
+                    public class GeneratedAttribute : global::System.Attribute
+                    {
+                    }
+                }
+                """, Encoding.UTF8));
+        });
+    }
+}
+```
+
+### Additional file transformation
+
+**User scenario:** As a generator author I want to be able to transform an external non-C# file into an equivalent C# representation.
+
+**Solution:** Use the `AdditionalTextsProvider` to filter for and retrieve your files. Transform them into the code you care about, then
+register that code with the solution.
+
+**Example:**
+
+```csharp
+[Generator]
+public class FileTransformGenerator : IIncrementalGenerator
+{
+    public void Initialize(IncrementalGeneratorInitializationContext context)
+    {
+        var pipeline = context.AdditionalTextsProvider.Select(static (text, cancellationToken) =>
+        {
+            if (!text.Path.EndsWith("*.xml"))
+            {
+                return null;
+            }
+
+            return (Name: text.Name, Code: MyXmlToCSharpCompiler.Compile(text.GetText(cancellationToken)));
+        })
+        .Where((pair, ct) => pair is not null);
+
+        context.RegisterSourceOutput(pipeline,
+            static (context, pair) => context.AddSource($"{pair.Name}generated.cs", SourceText.From(pair.Code, Encoding.UTF8)))
+    }
+}
+```
+
+### Augment user code
+
+**User scenario:** As a generator author I want to be able to inspect and augment a user's code with new functionality.
+
+**Solution:** Require the user to make the class you want to augment be a `partial class`, and mark it with a unique attribute.
+Provide that attribute in a `RegisterPostInitializationOutput` step. Register for callbacks on that attribute with
```

Comment:
Should we clarify that this approach is not required (in fact, it's not even recommended), and that for libraries that are also including a runtime library, the attributes should rather be defined there as part of the public API? This also matches what the BCL is doing, for instance, avoids duplicated attributes across projects and possibly conflicts with multiple projects with IVTs with each other, etc.

Reply:
> in fact, it's not even recommended

I'm not aware of such an anti-recommendation. I could certainly mention that a separate library is also an option though.

Reply:
@333fred The main issue here is `RegisterPostInitializationOutput` can't even access items defined in metadata references, so it's almost never usable in the context of IVTs. It ends up that one of the following two situations occurs:

1. The source generator is created for and only used by one project. In this case, `RegisterPostInitializationOutput` will work fine but provides no advantage over just adding the files directly to the project.
2. The source generator is used by potentially more than one project. In this case, `RegisterPostInitializationOutput` is insufficient for handling IVTs/metadata, and can't really be used.